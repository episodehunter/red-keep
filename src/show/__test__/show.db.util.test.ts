import { create } from 'chain-spy'
import {
  getShowIdFromDb,
  updateShow,
  findShow,
  filterOutNonExistingShows
} from '../show.db.util'

describe('Get show id', () => {
  test('Return show id', async () => {
    // Arrange
    const show = {
      tvdbId: 5
    }
    const dbRow = { id: 2 }
    const db = create({ where: () => Promise.resolve(dbRow) })

    // Act
    const result = await getShowIdFromDb(db, show)

    // Assert
    expect(result).toBe(2)
  })

  test('Reject if not exist', () => {
    // Arrange
    const show = {
      tvdbId: 5
    }
    const dbRow = null
    const db = create({ where: () => Promise.resolve(dbRow) })

    // Act and assert
    return expect(getShowIdFromDb(db, show)).rejects.toEqual(expect.any(Error))
  })

  test('Make same request every time', async () => {
    // Arrange
    const show = {
      tvdbId: 5
    }
    const dbRow = { id: 2 }
    const db = create({ where: () => Promise.resolve(dbRow) })

    // Act
    await getShowIdFromDb(db, show)

    expect(db.__execution_log__).toMatchSnapshot()
  })
})

test('Update show', async () => {
  // Arrange
  const show = {
    id: 5,
    tvdbId: 2,
    imdbId: 'tt12345',
    name: 'Dexter',
    airsDayOfWeek: 'Monday',
    airsTime: '09:00 PM',
    firstAired: '2016-03-15',
    genre: ['Drama', 'Action'],
    language: 'en',
    network: 'AMC',
    overview: 'Something, somthing',
    runtime: 60,
    ended: false,
    fanart: 'fanart.jpg',
    poster: 'poster.jpg',
    lastupdate: 1234567891
  }
  const db = create({ where: (key, value) => Promise.resolve(value) })

  // Act
  const result = await updateShow(db, 5, show)

  // Assert
  expect(result).toBe(5)
  expect(db.__execution_log__).toMatchSnapshot()
})

describe('Find show', () => {
  test('Resolve a show', async () => {
    // Arrange
    const dbShow = {
      id: 5,
      tvdb_id: 2,
      imdb_id: 'tt12345',
      name: 'Dexter',
      airs_dayOfWeek: 'Monday',
      airs_time: '09:00 PM',
      first_aired: '2016-03-15',
      genre: 'Drama|Action',
      language: 'en',
      network: 'AMC',
      overview: 'Something, somthing',
      runtime: '60',
      status: 'Ended',
      fanart: 'fanart.jpg',
      poster: 'poster.jpg',
      lastupdate: 1234567891
    }
    const db = create({ where: () => dbShow })

    // Act
    const result = await findShow(db, { id: 5 })

    // Assert
    expect(result).toMatchSnapshot()
    expect(db.__execution_log__).toMatchSnapshot()
  })

  test('Resolve null', async () => {
    // Arrange
    const dbShow = null
    const db = create({ where: () => dbShow })

    // Act
    const result = await findShow(db, { id: 5 })

    // Assert
    expect(result).toBe(null)
  })
})

test('Filter out non existing shows', async () => {
  // Arrange
  const db = create({ whereIn: () => [{ id: 2, tvdb_id: 5, imdb_id: 'Hello' }] })
  const ids = [0, 2, 5]

  // Act
  const result = await filterOutNonExistingShows(db, ids)

  // Assert
  expect(result).toEqual([{ id: 2, tvdbId: 5, imdbId: 'Hello' }])
  expect(db.__execution_log__).toMatchSnapshot()
})
