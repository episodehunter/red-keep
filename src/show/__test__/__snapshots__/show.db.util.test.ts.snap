// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Assert show exist Make same request every time 1`] = `
Array [
  Object {
    "name": "first",
    "type": "get",
  },
  Object {
    "args": Array [
      "id",
    ],
    "type": "apply",
  },
  Object {
    "name": "from",
    "type": "get",
  },
  Object {
    "args": Array [
      "tv_show",
    ],
    "type": "apply",
  },
  Object {
    "name": "where",
    "type": "get",
  },
  Object {
    "args": Array [
      Object {
        "tvdb_id": 5,
      },
    ],
    "type": "apply",
  },
]
`;

exports[`Assert show not exist Make same request every time 1`] = `
Array [
  Object {
    "name": "first",
    "type": "get",
  },
  Object {
    "args": Array [
      "id",
    ],
    "type": "apply",
  },
  Object {
    "name": "from",
    "type": "get",
  },
  Object {
    "args": Array [
      "tv_show",
    ],
    "type": "apply",
  },
  Object {
    "name": "where",
    "type": "get",
  },
  Object {
    "args": Array [
      Object {
        "tvdb_id": 5,
      },
    ],
    "type": "apply",
  },
]
`;

exports[`Filter out non existing shows 1`] = `
Array [
  Object {
    "name": "select",
    "type": "get",
  },
  Object {
    "args": Array [
      "id",
      "tvdb_id",
      "imdb_id",
    ],
    "type": "apply",
  },
  Object {
    "name": "from",
    "type": "get",
  },
  Object {
    "args": Array [
      "tv_show",
    ],
    "type": "apply",
  },
  Object {
    "name": "whereIn",
    "type": "get",
  },
  Object {
    "args": Array [
      "tvdb_id",
      Array [
        2,
        5,
      ],
    ],
    "type": "apply",
  },
]
`;

exports[`Find show Resolve a show 1`] = `
Object {
  "airsDayOfWeek": "Monday",
  "airsTime": "09:00 PM",
  "ended": true,
  "fanart": "fanart.jpg",
  "firstAired": "2016-03-15",
  "genre": Array [
    "Drama",
    "Action",
  ],
  "id": 5,
  "imdbId": "tt12345",
  "language": "en",
  "lastupdate": 1234567891,
  "name": "Dexter",
  "network": "AMC",
  "overview": "Something, somthing",
  "poster": "poster.jpg",
  "runtime": "60",
  "tvdbId": 2,
}
`;

exports[`Find show Resolve a show 2`] = `
Array [
  Object {
    "name": "first",
    "type": "get",
  },
  Object {
    "args": Array [
      "*",
    ],
    "type": "apply",
  },
  Object {
    "name": "from",
    "type": "get",
  },
  Object {
    "args": Array [
      "tv_show",
    ],
    "type": "apply",
  },
  Object {
    "name": "where",
    "type": "get",
  },
  Object {
    "args": Array [
      Object {
        "id": 5,
      },
    ],
    "type": "apply",
  },
]
`;

exports[`Update show 1`] = `
Array [
  Object {
    "args": Array [
      "tv_show",
    ],
    "type": "apply",
  },
  Object {
    "name": "update",
    "type": "get",
  },
  Object {
    "args": Array [
      Object {
        "airs_dayOfWeek": "Monday",
        "airs_time": "09:00 PM",
        "fanart": "fanart.jpg",
        "first_aired": "2016-03-15",
        "genre": "Drama|Action",
        "id": 5,
        "imdb_id": "tt12345",
        "language": "en",
        "lastupdate": 1234567891,
        "name": "Dexter",
        "network": "AMC",
        "overview": "Something, somthing",
        "poster": "poster.jpg",
        "runtime": 60,
        "status": "Continuing",
        "tvdb_id": 2,
      },
    ],
    "type": "apply",
  },
  Object {
    "name": "where",
    "type": "get",
  },
  Object {
    "args": Array [
      "id",
      5,
    ],
    "type": "apply",
  },
]
`;
